lang/java/java.py

```diff
@@ -158,14 +160,90 @@ def java_generic_data_structure(m) -> str:
     return public_camel_case_format_variable(m.text)
 
 
+class GenericTypeConnector(Enum):
+    AND = auto()
+    OF = auto()
+    DONE = auto()
+
+
+@mod.capture(rule="done")
+def java_generic_type_connector_done(m) -> GenericTypeConnector:
+    """Denotes ending a nested generic type"""
+    return GenericTypeConnector.DONE
+
+
+@mod.capture(rule="and|of|<user.java_generic_type_connector_done>")
+def java_generic_type_connector(m) -> GenericTypeConnector:
+    """Determines how to put generic type parameters together"""
+    with suppress(AttributeError):
+        return m.java_generic_type_connector_done
+    return GenericTypeConnector[m[0].upper()]
+
+
+@mod.capture(
+    rule="<user.java_generic_type_connector> <user.java_type_parameter_argument> [<user.java_generic_type_connector_done>]+"
+)
+def java_generic_type_continuation(m) -> list[Union[GenericTypeConnector, str]]:
+    """A generic type parameter that goes after the first using connectors"""
+    result = [m.java_generic_type_connector, m.java_type_parameter_argument]
+    with suppress(AttributeError):
+        dones = m.java_generic_type_connector_done_list
+        result.extend(dones)
+    return result
+
+
+@mod.capture(rule="<user.java_generic_type_continuation>+")
+def java_generic_type_additional_type_parameters(
+    m,
+) -> list[Union[GenericTypeConnector, str]]:
+    """Type parameters for a generic data structure after the first one"""
+    result = []
+    for continuation in m.java_generic_type_continuation_list:
+        result.extend(continuation)
+    return result
+
+
+def is_immediately_after_nesting_exit(pieces: list[str]) -> bool:
+    return len(pieces) >= 1 and pieces[-1] == ">"
+
+
+@mod.capture(
+    rule="<user.java_type_parameter_argument> [<user.java_generic_type_additional_type_parameters>]"
+)
+def java_type_parameter_arguments(m) -> str:
+    """Formatted Java type parameter arguments"""
+    parameters = [m.java_type_parameter_argument]
+    with suppress(AttributeError):
+        parameters.extend(m.java_generic_type_additional_type_parameters)
+    pieces = []
+    nesting: int = 0
+    for parameter in parameters:
+        if isinstance(parameter, str):
```

Comment:
I think you should be able to write this as `match parameter: case str: ...`. I'll test this and make the changes myself if it works!